deep assign/mixin through nested object keys

agubler opened the pull request:
**Type:** bug

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Addresses an issue with deep assigning and deep mixing objects where nested objects where not merged with target but overrode the target.

Resolves #290

agubler committed:
deep assign/mixin through nested object keys

bryanforbes commented on src/lang.ts:
```diff
@@ -40,36 +40,44 @@ interface MixinArgs<T extends {}, U extends {}> {
-				let value: any = (<any> source)[key];
+				let value: any = source[key];
+
+				if (copied.indexOf(value) !== -1) {
```
Won't this skip assignment if the same object is referenced by two properties of the source object?

```ts
const a = { foo: 1 };
const b = { one: a, two: a };
const c = deepAssign({}, b);
```

agubler replied:
@bryanforbes you are correct, I think I have now addressed this.

agubler committed:
ensure all refences in same object are copied over

agubler committed:
use the source not the value to track cyclical references.

agubler committed:
extend test

agubler commented:
@pottedmeat I believe this addresses the concerns we had and hopefully still guards against circular references. Please let me know if you can think of any more scenarios that are not covered by the unit tests.

agubler commented:
@kitsonk do you think you could give this a review? I believe it covers all the scenarios whilst also guarding against cyclical references.

kitsonk commented on src/lang.ts:
```diff
@@ -40,36 +40,45 @@ interface MixinArgs<T extends {}, U extends {}> {
 						value = copyArray(value, inherited);
 					}
 					else if (shouldDeepCopyObject(value)) {
+						const targetValue: any = target ? target[key] || {} : {};
```
appears there is only partial coverage on this line

kitsonk reviewed (commented):
Just one note.

agubler committed:
do not need to check target

kitsonk reviewed (approved):
👍
